src/webassembly/instance.rs

```diff
@@ -100,41 +92,62 @@ pub struct Instance {
     pub start_func: Option<FuncIndex>,
     // Region start memory location
     // code_base: *const (),
-
-    // C-like pointers to data (heaps, globals, tables)
-    pub data_pointers: DataPointers,
-
-    // Default memory bound
-    // TODO: Support for only one LinearMemory for now.
-    pub default_memory_bound: i32,
 }
 
 /// Contains pointers to data (heaps, globals, tables) needed
 /// by Cranelift.
+/// NOTE: Rearranging the fields will break the memory arrangement model
+#[repr(C)]
 #[derive(Debug)]
 pub struct DataPointers {
     // Pointer to tables
-    pub tables: UncheckedSlice<BoundedSlice<usize>>,
+    pub tables: TablesSlice,
 
     // Pointer to memories
-    pub memories: UncheckedSlice<UncheckedSlice<u8>>,
+    pub memories: MemoriesSlice,
 
     // Pointer to globals
-    pub globals: UncheckedSlice<u8>,
+    pub globals: GlobalsSlice,
+
 }
 
 impl Instance {
+    pub const TABLES_OFFSET: usize = 0; // 0 on 64-bit | 0 on 32-bit
+    pub const MEMORIES_OFFSET: usize = size_of::<TablesSlice>(); // 8 on 64-bit | 4 on 32-bit
+    pub const GLOBALS_OFFSET: usize = Instance::MEMORIES_OFFSET + size_of::<MemoriesSlice>(); // 16 on 64-bit | 8 on 32-bit
```

Comment:
Interesting, I was using a dynamic calculation in `module.rs` based on the pointer size in the latests commits of wasmer.
But I think I like yours better 💯 (and actually it's similar on how SpiderMonkey handles it).